onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -0,0 +1,476 @@
+          nnapi_model_->model_, op, input_indices.size(), &input_indices[0],
+          output_indices.size(), &output_indices[0]),
+      "op = " + std::to_string(op));
+}
+
+std::unique_ptr<Model> ModelBuilder::Compile() {
+  Prepare();
+
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksModel_identifyInputsAndOutputs(
+          nnapi_model_->model_, static_cast<uint32_t>(input_index_vec_.size()),
+          &input_index_vec_[0],
+          static_cast<uint32_t>(output_index_vec_.size()),
+          &output_index_vec_[0]),
+      "on identifyInputsAndOutputs");
+
+  if (use_fp16_) {
+    THROW_ON_ERROR_WITH_NOTE(
+        nnapi_->ANeuralNetworksModel_relaxComputationFloat32toFloat16(
+            nnapi_model_->model_, true),
+        "Set fp16");
+  }
+
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksModel_finish(nnapi_model_->model_),
+      "on model finish");
+
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksCompilation_create(nnapi_model_->model_, &nnapi_model_->compilation_),
+      "on create");
+
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksCompilation_setPreference(
+          nnapi_model_->compilation_, static_cast<int32_t>(exe_pref_)),
+      "on setPreference");
+
+  THROW_ON_ERROR_WITH_NOTE(
+      nnapi_->ANeuralNetworksCompilation_finish(nnapi_model_->compilation_),
+      "on compilation finish");
+
+  return std::move(nnapi_model_);
+}
+
+int32_t ModelBuilder::FindActivation(const std::string& output) {
+  int32_t fuse_code = ANEURALNETWORKS_FUSED_NONE;
+  const ONNX_NAMESPACE::NodeProto* activationNode{nullptr};
+  std::string node_name;
+  for (const auto& _node : model_proto_.graph().node()) {
+    if (_node.op_type() == "Relu" && output == _node.input(0)) {
+      fuse_code = ANEURALNETWORKS_FUSED_RELU;
+      activationNode = &_node;
+    }
+  }
+
+  if (fuse_code != ANEURALNETWORKS_FUSED_NONE) {
+    for (const auto& _node : model_proto_.graph().node()) {
+      if (&_node == activationNode)
+        continue;
+
+      // if there is any other node using the output
+      // will add relu separately
+      for (const auto& node_input : _node.input()) {
+        if (output == node_input)
+          return ANEURALNETWORKS_FUSED_NONE;
+      }
+    }
+
+    // if output is a graph output
+    // will add relu separately
+    for (const auto& model_output : model_proto_.graph().output()) {
+      if (model_output.name() == output)
+        return ANEURALNETWORKS_FUSED_NONE;
+    }
+
+    fused_activations_.insert(activationNode->name());
+  }
+
+  return fuse_code;
+}
+
+IOpBuilder* ModelBuilder::GetOpBuilder(const ONNX_NAMESPACE::NodeProto& node) {
+  if (!Contains(op_builders_, node.op_type()))
+    return nullptr;
+
+  return op_builders_[node.op_type()].get();
+}
+
+std::string ModelBuilder::GetUniqueName(const std::string& base_name) {
+  std::ostringstream os;
+  os << base_name << "_token_" << name_token_++;
+  return os.str();
+}
```

Comment:
Is there any sort of name registry we can check to ensure it's unique. Granted the odds of a clash are very low, but it may be hard to debug if it happens.